relentless/simulate/lammps.py

```diff
@@ -0,0 +1,347 @@
+        else:
+            cmds += ['region box block {} {} {} {} {} {}'.format(*box[:-3])]
+        cmds += ['create_box {N} box'.format(N=len(sim.ensemble.types))]
+
+        # use lammps random initialization routines
+        for i in sim.ensemble.types:
+            if sim.ensemble.N[i] is None:
+                raise ValueError('Number of particles for type {} must be set.'.format(i))
+
+            cmds += ['create_atoms {typeid} random {N} {seed} box'.format(typeid=sim.type_map[i],
+                                                                          N=sim.ensemble.N[i],
+                                                                          seed=self.seed+sim.type_map[i]-1)]
+        cmds += ['mass * 1.0',
+                 'velocity all create {temp} {seed}'.format(temp=sim.ensemble.T,
+                                                            seed=self.seed)]
+
+        cmds += self.attach_potentials(sim)
+
+        return cmds
+
+class MinimizeEnergy(LAMMPSOperation):
+    def __init__(self, energy_tolerance, force_tolerance, max_iterations, dt):
+        self.energy_tolerance = energy_tolerance
+        self.force_tolerance = force_tolerance
+        self.max_iterations = max_iterations
+
+    def to_commands(self, sim):
+        cmds = ['minimize {etol} {ftol} {maxiter} {maxeval}'.format(etol=self.energy_tolerance,
+                                                                    ftol=self.force_tolerance,
+                                                                    maxiter=self.max_iterations,
+                                                                    maxeval=100*self.max_iterations)]
+
+        return cmds
+
+class AddLangevinIntegrator(LAMMPSOperation):
+    def __init__(self, dt, friction, seed):
+        self.friction = friction
+        self.seed = seed
+
+        self._fix_langevin = self.new_fix_id()
+        self._fix_nve = self.new_fix_id()
+
+    def to_commands(self, sim):
+        # obtain per-type mass
+        Ntypes = len(sim.ensemble.types)
+        _mass = sim.lammps.numpy.extract_atom('mass')
+        if _mass is None or _mass.shape != (Ntypes+1,1):
+            raise ValueError('Per-type masses not set.')
+        mass = np.array([i[0] for i in _mass[1:]])
+
+        # obtain per-type friction factor
+        friction = np.zeros_like(mass)
+        for t in sim.ensemble.types:
+            try:
+                friction[sim.type_map[t]-1] = self.friction[t]
+            except TypeError:
+                friction[sim.type_map[t]-1] = self.friction
+            except KeyError:
+                return KeyError('The friction factor types must match the atom types.')
+
+        # compute per-type damping parameter
+        damp = np.divide(mass, friction, where=(friction>0))
+        damp_ref = damp[0]
+        scale = damp/damp_ref
+        scale_str = ' '.join(['scale {} {}'.format(i+1,s) for i,s in enumerate(scale[1:])])
```

Comment:
This is changing the lammps 1-indexing back to 0-indexing. Since this is only used internally, I think it would be more readable to stick with the LAMMPS indexing and make a note about it.
```suggestion
        # obtain per-type mass
        # all arrays will be 1-indexed using the lammps convention for typeid
        Ntypes = len(sim.ensemble.types)
        mass = sim.lammps.numpy.extract_atom('mass')
        if mass is None or mass.shape != (Ntypes+1,1):
            raise ValueError('Per-type masses not set.')

        # obtain per-type friction factor
        friction = np.zeros_like(mass)
        for t in sim.ensemble.types:
            try:
                friction[sim.type_map[t]] = self.friction[t]
            except TypeError:
                friction[sim.type_map[t]] = self.friction

        # compute per-type damping parameter
        damp = np.divide(mass, friction, where=(friction>0))
        damp_ref = damp[1]

        # rescale damp parameters if there are multiple types
        if Ntypes > 1:
            scale = damp/damp_ref
            scale_str = ' '.join(['scale {} {}'.format(i+1,s) for i,s in enumerate(scale[2:])])
        else:
            scale_str = ''
```

I also made the following changes:
1. Remove the `KeyError`. I'm not sure I see the case where this would ever trigger? Do you have an example?
2. Only compute scale if there are at least 2 types, as this would produce an error with a 1-type system.